pipeline_dp/dataset_histograms/sum_histogram_computation.py

```diff
@@ -0,0 +1,350 @@
+
+    # (index, [hist.FrequencyBin])
+
+    def bins_to_histogram(bins):
+        sorted_bins = sorted(bins, key=lambda bin: bin.lower)
+        return hist.Histogram(name, sorted_bins)
+
+    col = backend.map_values(col, bins_to_histogram, "To histogram")
+
+    col = backend.to_list(col, "To 1 element collection")
+
+    def sort_histograms_by_index(index_histogram):
+        if len(index_histogram) == 1:
+            # It is a histogram for one column, return it w/o putting it in a list.
+            return index_histogram[0][1]
+
+        # Sort histograms by index and return list of them.
+        # Beam does not like mutating arguments, so copy the argument.
+        index_histogram = copy.deepcopy(index_histogram)
+        return [histogram for index, histogram in sorted(index_histogram)]
+
+    col = backend.map(col, sort_histograms_by_index, "sort histogram by index")
+    # 1 element collection with list of histograms: [hist.FrequencyBin]
+    return col
+
+
+def _create_bucket_generators_per_key(
+        col, number_of_buckets: int, backend: pipeline_backend.PipelineBackend):
+    """Creates bucket generators per key.
+
+    Args:
+        col: collection of (key, value)
+        backend: PipelineBackend to run operations on the collection.
+        num_buckets: the number of buckets in the output histogram.
+
+    Returns:
+        1 element collection with dictionary {key: LowerUpperGenerator}.
+    """
+    col = pipeline_functions.min_max_per_key(
+        backend, col, "Min and max value per value column")
+    # (index, (min, max))
+
+    col = backend.to_list(col, "To list")
+
+    # 1 elem col: ([(index, (min, max))])
+
+    def create_generators(index_min_max: List[Tuple[int, Tuple[float, float]]]):
+        min_max_sorted_by_index = [v for k, v in sorted(index_min_max)]
+        return [
+            LowerUpperGenerator(min, max, number_of_buckets)
+            for min, max in min_max_sorted_by_index
+        ]
+
+    return backend.map(col, create_generators, "Create generators")
+
+
+def _flat_values(col, backend: pipeline_backend.PipelineBackend):
+    """Unnests values in (key, value) collection.
+
+    Args:
+        col: collection of (key, value) or (key, [value])
+        backend: PipelineBackend to run operations on the collection.
+
+    Transform each element:
+    (key, value: float) -> ((0, key), value)
+    (key, value: list[float]) -> [((0, key), value[0]), ((1, key), value[1])...]
+    and then unnest them.
+
+    Return:
+        Collection of ((index, key), value).
+    """
+
+    def flat_values(key_values):
+        key, values = key_values
+        if isinstance(values, Iterable):
+            for i, value in enumerate(values):
+                yield (i, key), value
+        else:
+            yield (0, key), values  # 1 value
+
+    return backend.flat_map(col, flat_values, "Flat values")
+
+
+def _compute_linf_sum_contributions_histogram(
```

Comment:
All the functions below these seem to be doing very similar things. Do you think it would be possible to merge them into a single function with a parameter for choosing the type of histogram? Would that be more readable?

Reply:
All of those functions have different input formats, all they have format:
```
<pre-processing>
_compute_frequency_histogram_per_key
```

And all heavy lifiting has been already extracted into `_compute_frequency_histogram_per_key`, and pre-processing is simple.